cloudinit/distros/__init__.py

```diff
@@ -818,20 +823,108 @@ def create_user(self, name, **kwargs):
             return self.add_snap_user(name, **kwargs)
 
         # Add the user
-        self.add_user(name, **kwargs)
-
-        # Set password if plain-text password provided and non-empty
-        if "plain_text_passwd" in kwargs and kwargs["plain_text_passwd"]:
-            self.set_passwd(name, kwargs["plain_text_passwd"])
-
-        # Set password if hashed password is provided and non-empty
-        if "hashed_passwd" in kwargs and kwargs["hashed_passwd"]:
-            self.set_passwd(name, kwargs["hashed_passwd"], hashed=True)
+        pre_existing_user = not self.add_user(name, **kwargs)
+
+        blank_password = False
+        has_password = False
+        password_key = None
+
+        if "plain_text_passwd" in kwargs:
+            has_password = True
+            password_key = "plain_text_passwd"
+            if kwargs["plain_text_passwd"]:
+                # Set password if plain-text password provided and non-empty
+                self.set_passwd(name, kwargs["plain_text_passwd"])
+            else:
+                blank_password = True
+
+        if "hashed_passwd" in kwargs:
+            has_password = True
+            password_key = "hashed_passwd"
+            if kwargs["hashed_passwd"]:
+                # Set password if hashed password is provided and non-empty
+                self.set_passwd(name, kwargs["hashed_passwd"], hashed=True)
+            else:
+                blank_password = True
+
+        if pre_existing_user:
+            if not has_password:
+                if "passwd" in kwargs:
+                    password_key = "passwd"
+                    # Only "plain_text_passwd" and "hashed_passwd"
+                    # are valid for an existing user.
+                    LOG.warning(
+                        "'passwd' in user-data is ignored for existing "
+                        "user %s",
+                        name,
+                    )
 
-        # Default locking down the account.  'lock_passwd' defaults to True.
-        # lock account unless lock_password is False.
+                # As no password set for the existing user in user-data then
+                # check if the existing user's hashed password value in
+                # /etc/shadow is blank (whether locked or not).
+                cmd = [
+                    "grep",
+                    "-q",
+                    "-e",
+                    "^%s::" % name,
+                    "-e",
+                    "^%s:!:" % name,
+                    "/etc/shadow",
```

Comment:
There are some enviroments (util.is_snappy()) where /etc/shadow is read-only and any added users by cloud-init or other are forced to use `--extrausers` flag to store those users in /var/lib/extrausers/shadow. We may want to grep that file too if util.is_snappy()

Reply:
Hardcoding a file path like `/etc/shadow` here is probably better places as an attr on the class or instance such as `shadow_fns: [...]` which gets  is queried based on the distro you are running in. In most cases I expect this to just be a one-item list, but in snappy case it'd be two files to process/parse/grep

Reply:
May be worth considering using `passwd --status <user>`

$ sudo passwd --status <user> | awk '{print $2}'
P  => password
NP => no password
L => locked

Reply:
Excellent suggestion @cjp256! that does take into account both /etc/shadow and /var/lib/extrausers sources. It'd be better to rely on tooling than a grep as well.

Reply:
The question remains, is `passwd --status` supported broadly in most distros.

Reply:
> May be worth considering using `passwd --status <user>`

Nope, that cannot distinguish between a password field value of "!" and "!$6$..." (in both cases it indicates "L") which is **exactly** what the grep is intended to do.

Reply:
> There are some enviroments (util.system_is_snappy()) where /etc/shadow is read-only and any added users by cloud-init or other are forced to use `--extrausers` flag to store those users in /var/lib/extrausers/shadow. We may want to grep that file too if util.system_is_snappy()

Ah ok, I know nothing about snappy.

So with snappy should the /var/lib/extrausers/shadow file be grepped first and then /etc/shadow only grepped if the user has no entry at all in the /var/lib/extrausers/shadow file?

Reply:
@dermotbradley yes. if /var/lib/extrausers/shadow exists we can grep that first and fallback to /etc/shadow if the user has no entry at all in /var/lib/extrausers/shadow +1

Reply:
@blackboxsw done.